problemtools/generatedata.py

```diff
@@ -0,0 +1,272 @@
+                        type=int,
+                        default=None,
+                        help='level of parallelism')
+    argparser_basic_arguments(parser)
+    parser.add_argument('problemdir', nargs='+')
+    return parser
+
+
+def clean(prob, args):
+    ProblemAspect.errors = 0
+    ProblemAspect.warnings = 0
+    base_path = os.path.join(prob.probdir, 'data')
+
+    testcases = {
+        case['path']: case
+        for case in prob.generators._testcases
+    }
+
+    def walk(name, path):
+        case_count = 0
+        cases = set()
+        empty = True
+        for fname in sorted(os.listdir(path)):
+            curpath = os.path.join(path, fname)
+            nice_path = os.path.relpath(curpath, base_path)
+            if '.' in fname:
+                fname, ext = fname.split('.', 1)
+            else:
+                ext = ''
+            curname = '%s/%s' % (name, fname)
+
+            if os.path.isdir(curpath):
+                next_empty, next_cases = walk(curname, curpath)
+                case_count += next_cases
+                if next_empty:
+                    if not args.dry_run:
+                        os.rmdir(curpath)
+                else:
+                    empty = False
+            else:
+                remove = args.clean_all
+                is_case = False
+                if (fname, ext) == ('testdata', 'yaml'):
+                    if curname + '.yaml' in prob.generators._testdata_yaml:
+                        remove = True
+                elif curname in testcases:
+                    is_case = True
+                    case = testcases[curname]
+                    if ext == 'in':
+                        remove = not case['manual']
+                    elif ext == 'ans':
+                        remove = case['solution'] is not None
+                    elif ext in Generators._VISUALIZER_EXTENSIONS:
+                        remove = case['visualizer'] is not None
+
+                if remove:
+                    prob.generators.msg('Removing %s' % nice_path)
+                    if not args.dry_run:
+                        os.unlink(curpath)
+                    if is_case and curname not in cases:
+                        cases.add(curname)
+                        case_count += 1
+                else:
+                    empty = False
+
+        return empty, case_count
+
+    cases_cleaned = 0
+    for directory in prob.generators._data_directories:
+        path = os.path.join(base_path, directory)
+        if os.path.isdir(path):
+            cases_cleaned += walk('data/%s' % directory, path)[1]
+    return cases_cleaned, ProblemAspect.errors, ProblemAspect.warnings
+
+
+class GenerateState:
+    prob = None
+    args = None
+
+def generate_case(case_idx):
+    ProblemAspect.errors = 0
+    ProblemAspect.warnings = 0
+    prob = GenerateState.prob
+    args = GenerateState.args
+    case = prob.generators._testcases[case_idx]
+
+    steps = [
```

Comment:
oh this is nice! Maybe I can steal this idea to simplify some code ;)

Reply:
Do you run input/output validators on the generated cases?

Reply:
No, that's left as the responsibility of the `verifyproblem` command.